build_tools/cmake/build_and_test_asan.sh

```diff
@@ -0,0 +1,151 @@
+#!/bin/bash
+
+# Copyright 2020 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Note: this script diverges from the non-ASan build in a few ways:
+#   * The CMake build sets `IREE_ENABLE_ASAN=ON`
+#   * Omit optional components that don't work with ASan (e.g. Python bindings)
+#   * Some tests that fail under ASan are individually excluded
+#
+# The desired build directory can be passed as
+# the first argument. Otherwise, it uses the environment variable
+# IREE_TSAN_BUILD_DIR, defaulting to "build-tsan". Designed for CI, but
+# can be run manually. This reuses the build directory if it already exists.
+#
+# Build and test the project with CMake with ASan enabled and using
+# SwiftShader's software Vulkan driver.
+# ASan docs: https://clang.llvm.org/docs/AddressSanitizer.html
+
+
+set -xeuo pipefail
+
+ROOT_DIR="${ROOT_DIR:-$(git rev-parse --show-toplevel)}"
+cd "${ROOT_DIR}"
+
+CMAKE_BIN=${CMAKE_BIN:-$(which cmake)}
+BUILD_DIR="${1:-${IREE_TSAN_BUILD_DIR:-build-asan}}"
+IREE_ENABLE_ASSERTIONS="${IREE_ENABLE_ASSERTIONS:-ON}"
+IREE_ENABLE_CCACHE="${IREE_ENABLE_CCACHE:-OFF}"
+
+"$CMAKE_BIN" --version
+ninja --version
+
+if [[ -d "${BUILD_DIR}" ]]; then
+  echo "Build directory '${BUILD_DIR}' already exists. Will use cached results there."
+else
+  echo "Build directory '${BUILD_DIR}' does not already exist. Creating a new one."
+  mkdir "${BUILD_DIR}"
+fi
```

Comment:
Do we still need this on the new CI system?

Reply:
This is copying from the tsan build script and the other build scripts do this as well, so I think it's good to keep for consistency. IIRC, the main motivation here was to make local testing of the build scripts easier and it doesn't hurt for CI